### PR TITLE
Allow ?-converting from Result<T, E> in functions returning Option<Result<T, E>>

### PR DESCRIPTION
**Pull request by GoldsteinE**
This PR adds implementation `FromResidual<Result<Infallible, E>> for Option<Result<T, E>>`, allowing to use `?` operator on `Result<_, E>` in functions returning `Option<Result<T, E>>`.

Interestingly, this is already the case for `Poll<Option<Result<T, E>>>`, so this implementation seems to be accidentally missing and not deliberately left out.

This is particularly useful for `Iterator` implementations where `Item = Result<T, E>`.

This implementation is `const` when underlying `From<_>` impl is `const`. This is not the case for `Poll<_>` implementations and I’m not sure why. Should I add missing `const` to `Poll<_>` impls or there’s a reason for that?

### CODE REVIEW

**Comment by GoldsteinE** [comment]
I’m not sure if this needs an API change proposal, but I’m happy to create one if needed.

**Comment by joshtriplett** [comment]
The `Poll` implementations of `?` aren't something to take inspiration from; we've had a fair bit of trouble with those, with the benefit of hindsight. cc @scottmcm

Nonetheless, this *does* seem potentially useful.

The main downside I could imagine would be a context in which *some* errors need to get processed in a different way and then turned into `None`. (For instance, failures that will prevent iterating at all; if those get turned into `Some(Err(...))` then that could result in an infinite loop.)

**Comment by GoldsteinE** [comment]
> The Poll implementations of ? aren't something to take inspiration from; we've had a fair bit of trouble with those, with the benefit of hindsight. cc @scottmcm

I think that even if `Poll<_>` implementations were a mistake, it’s surprising to have implementation for `Poll<Option<Result<_, _>>>` and not for `Option<Result<_, _>>`.

> The main downside I could imagine would be a context in which some errors need to get processed in a different way and then turned into None. (For instance, failures that will prevent iterating at all; if those get turned into Some(Err(...)) then that could result in an infinite loop.)

I thought about it. I think this conversion as implemented follows the general pattern “implicit conversions do not lose data”. It’s also very simple to go from `Result<_, _>` to `Option<_>` with `.ok()`, but returning an error requires something cumbersome like

```rust
match res {
    Ok(val) => val,
    Err(err) => return Some(Err(err)),
}
```
and can’t be abstracted into a method.

**Comment by scottmcm** [comment]
Convenient that I got randomly picked for this review 🙃

I was going to say the same thing that Josh did -- that I don't think experience has been kind to these complex implementations on `Poll`.  (cc @yaahc, who might have more to say on this.)  Note also that trait implementations are insta-stable, so we can't really experiment with them.  That generally means that they can't be accepted without a "they seem pretty obviously what we want" sentiment.  (Things like new inherent methods, where `#[unstable]` works, are much easier to accept provisionally and see how they go.)

I think the core problem is that the nestings don't permute well.  `?` on an `Option<Result<T, E>>` in a function returning `Result<T, E>` isn't going to ever work without violating the "the value produced by `?` doesn't depend on the context" rule that https://github.com/rust-lang/rfcs/pull/1859 decided to have which, AFAIK, is still popular.  So if the other way around doesn't work, it's not clear to me that this way 'round should be supported.

---

> This is particularly useful for `Iterator` implementations where `Item = Result<T, E>`.

Hmm, interesting.  (And coincidentally reminds me of https://swatinem.de/blog/fallible-iterators/ from this week's TWiR).

Can you show some examples of the code you're thinking of using this?  I'd like to experiment with other possibilities, like whether `try{}` blocks would help.  (If not, that'd be particularly good information for the error handling story.)

**Comment by GoldsteinE** [comment]
I don’t have public examples off the top of my head, but quick rg in `rust-lang/rust` shows a couple of places where this could be used: [std::io::Lines](https://github.com/rust-lang/rust/blob/44dbd9808e872f52b83c9238f18046bb73df330a/library/std/src/io/mod.rs#L2810) and [std::io::Split](https://github.com/rust-lang/rust/blob/44dbd9808e872f52b83c9238f18046bb73df330a/library/std/src/io/mod.rs#L2779). They are not very interesting though, since it’s only a single usage.

GitHub search returns some more results:
* https://github.com/nix-rust/nix/blob/5f859d10dfd21b99405054d80392649adc392fd0/src/dir.rs#L97
* https://github.com/m-ou-se/gpio-cdev/blob/bd58cf1a8f1438c0597de23b1b5823342dc48412/src/lib.rs#L179 and also lower on line 1017
* https://github.com/tmerr/i3ipc-rs/blob/0a06e74d7a4ab448b081ce897e15c0cb273ba56d/src/lib.rs#L233 — this one could be simplified significantly (though it’d require adding `From<_>` impls for the error type).

>  So if the other way around doesn't work, it's not clear to me that this way 'round should be supported.

I’d like to note that `?` is already asymmetric, since it relies on `From<_>` to convert errors.

**Comment by GoldsteinE** [comment]
@joshtriplett I noticed you’ve added “needs FCP” label. Should I write one?

**Comment by c410-f3r** [comment]
Such implementation is ***SO*** useful when you are dealing with fallible iterators that return both `Err`s and `Ok`s.

```rust
use core::num::TryFromIntError;
let array: [Result<i16, TryFromIntError>; 4] = [Ok(1), Ok(-256), Ok(3), Ok(4)];
let mut iter = array.into_iter().filter_map(|rslt| {
    let elem = match rslt {
        Err(err) => return Some(Err(err)),
        Ok(elem) => elem
    };
    let _u8 = match u8::try_from(elem) {
        Err(err) => return Some(Err(err)),
        Ok(elem) => elem
    };
    (_u8 % 2 == 0).then_some(Ok::<_, TryFromIntError>(_u8))
});
assert!(matches!(iter.next().unwrap(), Err(_)));
assert!(matches!(iter.next().unwrap(), Ok(4)));
```

Just an example but there can be other combinators that involve `Option`. As seen above, it is **very** cumbersome to always manually `match` every `Result`.

```rust
use core::num::TryFromIntError;
let array: [Result<i16, TryFromIntError>; 4] = [Ok(1), Ok(-256), Ok(3), Ok(4)];
let mut iter = array.into_iter().filter_map(|rslt| {
    let _u8 = u8::try_from(rslt?)?;
    (_u8 % 2 == 0).then_some(Ok::<_, TryFromIntError>(_u8))
});
assert!(matches!(iter.next().unwrap(), Err(_)));
assert!(matches!(iter.next().unwrap(), Ok(4)));
```
Taking aside macros, I am not aware of any other thing that improves the first snippet, therefore, this PR seems worthwhile.

**Comment by Numenorean** [comment]
I have recenlty faced with the same issue while it's more convenient to use built-in implementation instead of writing the same code each time. Having such need as writing plenty of iterators (in my case it's scrapers) I obviously want to use fallible iterators because there is no other way to throw errors away. Sorry for not reproducible examples, it's just from the my current project
```rust
let mut owners = asset.into_owners_crawler(self.graphql_client).into_iter();
let owner = match owners.next() {
    Some(Ok(owner)) => match self.database.owner_exists_or_add(&owner, &asset_id) {
        Ok(false) => owner,
        Ok(true) => return Some(Err(ScrapeError::OwnerRepeat)),
        Err(err) => return Some(Err(err.into())),
    },
    Some(Err(error)) => return Some(Err(error)),
    None => unreachable!("owners seems to be empty, should not happened: {asset_id}"),
};
```
And after implementing `Try` on `Result<_, _>` in functions returning `Option<Result<_, _>>` 

```rust
let mut owners = asset.into_owners_crawler(self.graphql_client).into_iter();
let owner = match owners.next().transpose()? {
    None => unreachable!("owners seems to be empty, should not happened: {asset_id}"),
    Some(owner) => {
        if self.database.owner_exists_or_add(&owner, &asset_id)? {
            return Some(Err(ScrapeError::OwnerRepeat));
        }
        owner
    }
};
```
Modified version seems to be more readable and convinient to work with. Moreover after modifing `owner_exists_or_add` so it returns just a result we have got even shorter logic

```rust
let mut owners = asset.into_owners_crawler(self.graphql_client).into_iter();
let owner = match owners.next().transpose()? {
    None => unreachable!("owners seems to be empty, should not happened: {asset_id}"),
    Some(owner) => self.database.add_owner(&owner, &asset_id)?,
};
```
At this stage we have got rid of all unwanted pattern matching. If our codebase built on top of iterators we write lots of unnecessary code, though, in a way, it could be solved with macros

**Comment by m-ou-se** [comment]
> The `Poll` implementations of `?` aren't something to take inspiration from; we've had a fair bit of trouble with those, with the benefit of hindsight.

The implementation proposed in this PR is less problematic though. The issue with the controversial `Poll` implementations was that it's not obvious whether a `?` applied to a `Poll<Result>` expression applies to the `Poll` or the `Result`. The implementation in this PR is just about applying `?` to a `Result`, so there's no ambiguity. The `Option<Result>` is the return type of the function, not of the expression that the `?` gets applied to.

**Comment by m-ou-se** [comment]
I am wondering if this implementation would be too specific, and if it'd make sense to try to find a way to be more generic. Probably not, because a more generic version would also apply to `Option<Option>`, which conflicts with the existing implementation.

**Comment by yaahc** [comment]
> I think that even if `Poll<_>` implementations were a mistake, it’s surprising to have implementation for `Poll<Option<Result<_, _>>>` but not for `Option<Result<_, _>>`.

An unfortunate price to pay for stability. We already have inconsistencies like this such as how `Box<dyn Error>` does not impl `Error` but `Arc<dyn Error>` does. We tend to prefer inconsistencies over repeating mistakes. That said, I'm not positive the `Poll` impls were a mistake, I know some people swear by them, and I'm not in the habit of writing manual impls of `Future` so I haven't had much occasion to use them.

---

Replying to both @scottmcm and @c410-f3r, here's a version of the example @c410-f3r gave using try blocks like @scottmcm mentioned:

```rust
use core::num::TryFromIntError;
let array: [Result<i16, TryFromIntError>; 4] = [Ok(1), Ok(-256), Ok(3), Ok(4)];
let mut iter = array
    .into_iter()
    .filter_map(|rslt| -> Option<Result<_, TryFromIntError>> {
        try {
            try {
                let elem = rslt?;
                let _u8 = u8::try_from(elem)?;
                if _u8 % 2 != 0 {
                    return None;
                }
                _u8
            }
        }
    });
assert!(matches!(iter.next().unwrap(), Err(_)));
assert!(matches!(iter.next().unwrap(), Ok(4)));
```

vs

```rust
use core::num::TryFromIntError;
let array: [Result<i16, TryFromIntError>; 4] = [Ok(1), Ok(-256), Ok(3), Ok(4)];
let mut iter = array.into_iter().filter_map(|rslt| {
    let _u8 = u8::try_from(rslt?)?;
    (_u8 % 2 == 0).then_some(Ok::<_, TryFromIntError>(_u8))
});
assert!(matches!(iter.next().unwrap(), Err(_)));
assert!(matches!(iter.next().unwrap(), Ok(4)));
```

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=54f9c428da373dcc993adde3c92aef5c

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/libs-api meeting. Based on @yaahc's suggested use of `try`, we generally felt that this wouldn't add as much value as it would add the chance for confusion. Nonetheless, we're leaving this open for further discussion to see if there are further use cases that `try` wouldn't help with.

**Comment by scottmcm** [comment]
Hmm, the `try { try {` doesn't fill me with joy.

Another version of this would use `transpose`:

```rust
use core::num::TryFromIntError;
let array: [Result<i16, TryFromIntError>; 4] = [Ok(1), Ok(-256), Ok(3), Ok(4)];
let mut iter = array
    .into_iter()
    .filter_map(|rslt| -> Option<Result<_, TryFromIntError>> {
        let x: Result<_, _> = try {
            let u8 = u8::try_from(rslt?)?;
            (u8 % 2 == 0).then_some(u8)
        };
        x.transpose()
    });
assert!(matches!(iter.next().unwrap(), Err(_)));
assert!(matches!(iter.next().unwrap(), Ok(4)));
```

And, helpfully, if we get https://github.com/rust-lang/rust/pull/98417 then I think that could simplify down to
```rust
let mut iter = array
    .into_iter()
    .filter_map(|rslt| {
        try {
            let u8 = u8::try_from(rslt?)?;
            (u8 % 2 == 0).then_some(u8)
        }.transpose()
    });
```

which might be sufficient

**Comment by yaahc** [comment]
Independent of the `try` blocks issue, I believe we should do a crater run if we do decide to proceed with this. My primary concerns at the moment are:

* breaking existing inference
* compiler diagnostic regressions
* locking us out of future try impls we may prefer
* allowing code that previously didn't compile to compile in unexpected ways.

**Comment by rkuhn** [comment]
Following [some discussion](https://internals.rust-lang.org/t/working-with-fallible-iterators/17136) here are my 2c:

- the one-line idiom of “bind success value to local name, propagate error outwards” using `?` has been very successful in making error-aware code readable
- the above suggestions of using `try` blocks do not achieve this for the specific use-case of implementing or working with a fallible iterator — the point is to get rid of the visual clutter and focus on the logic (indentation depth of nested `try` also is a problem)
- the proposed `impl FromResidual` fixes this nicely and in a neatly targeted way

As @scottmcm notes, such an addition cannot be tried out in unstable form, which raises the bar for merging this PR. While I cannot judge the points raised by @yaahc concerning compiler internals or the team’s future plans, I am pretty sure that supporting `?` where it is currently rejected shouldn’t break existing code.

I concur with the final point and would rephrase it as “`?` should only do the obvious thing, no magic”, so code shouldn’t compile in unexpected ways. With the proposed `impl` we are looking at the following patterns:

```rust
fn f(v: Vec<u32>) -> Option<Result<(), Error>> {
    let x = try_something()?; // turns Err into Some(Err)
    let y = v.last()?; // propagates None
    let z = u8::try_from(y).ok()?; // turns Err into None
    ...
}
```

Due to the specificity of the `impl` these cases are all very obvious, as can easily be understood by reading the resulting docs. Making the `impl` more generic is bound to ruin this property, and I have trouble understanding along what axis parametericity could be added: `Try<Try<...>>` would have unclear semantics as pointed out above (if I understand correctly), `Option<Option<...>>` can already be flattened, as can `Result<Result<...>>`.

In closing, what I’m after is the ability to express a simple and single step of my logic on a single line if possible, which currently is only possible with custom macros. `try` does not seem to fix this as far as I have been able to determine.

**Comment by simonbuchan** [comment]
Possibly dumb alternative:

```rust
trait TryIterator {
  type Item;
  fn try_next(&mut self) -> Result<Option<Self::Item>>;
}

impl<T> Iterator for T
where T: TryIterator {
  type Item = T::Item;
  fn next(&mut self) -> Option<Result<Self::Item>> {
    self.try_next().transpose()
  }
}
```

If I understand the request, this means you could instead implement TryIterator and `?` then did what you want, and less confusingly to me?

**Comment by rkuhn** [comment]
That approach just inverts the problem: instead of decorum for bailing on `Err` I now need decorum for bailing on `None`. It also only fixes implementing an iterator, but not working with it (like `.filter_map`).

It feels a bit weird to add a generic mechanism like Try/FromResidual/Residual to the standard library and then implement it for a very limited but inconsistent set of cases — if `?` was just meant to be used with plain `Option` and `Result` then why not make it a special language feature? (I think I know why, and I’d fully agree: language features should be as generic and composable as possible, to get the most benefit out of occupying a slot in the complexity budget — hence my above post.)

**Comment by simonbuchan** [comment]
Fair enough, but I'm guessing at least wanting `None?` would be a lot less common?

You could also add `try_* ` methods, such as `try_filter_map`, to `Iterator<Item=Result>` with the same effect.

**Comment by rkuhn** [comment]
> Fair enough, but I'm guessing at least wanting None? would be a lot less common?

While “common” is highly subjective, the structure of the return type is not: `Option<Result<...>>` or `Result<Option<...>>` inherently have two failure modes, one with no value and one with an error value. Since these have distinct types, it is reasonable to be able to yield either one when bailing out.

> You could also add try_* methods, such as try_filter_map, to Iterator<Item=Result> with the same effect.

Yes, with the same effect as your TryIterator proposal, which does not solve but only invert the problem. Plus I cannot add these methods to the standard library — and the whole point of this discussion is removing clutter from error-aware code without having to maintain extras (like macros, extension traits, etc.).

**Comment by simonbuchan** [comment]
Yep, exactly. But part of design is checking if alternatives work at least well enough, and adding `Result`-friendly iterator traits and methods that can be marked `#[unsafe]` seems reasonable to check, at least.

**Comment by scottmcm** [comment]
Since this needs a libs-api decision (& possible FCP), it being assigned to me isn't helpful, since I'm not a team member.  As such,
r? libs

**Comment by yaahc** [comment]
I don't want to reject this outright, but I do think we should hold off on possibly adding this until after try blocks have been stabilized for a while. Then we will be in a much better position to judge how necessary this impl is.

**Comment by rkuhn** [comment]
Thanks for those concerns, if there is indeed a silent refactoring failure hidden inside this addition then I agree we shouldn’t do it.

As far as I see it, and assuming no change to the function body:

- if we start with `fn(...)->Result<T, E>` then all `?` will now propagate `Some(Err(T))`
- if we start with `fn(...)->Option<Result<T, E>>` then `?` on `Result` will still do what it did before and on `Option` it will fail to compile

Only the first case is a potential problem. I’d argue that due to `Option<Result<T, E>>` being isomorphic to `Result<Option<T>, E>` the behaviour is not actually changed. The desired effect of the refactoring may well be that some errors should be turned into None instead, leaving others alone (otherwise there would be no need for Result anymore). This is a manual process in any case, where each potential break propagation needs to be inspected — it can’t be driven by types since there is only a single `E` type. The argument against this proposed addition then is that flagging all `?` as type errors after the refactoring will ensure that all are inspected by the programmer; after to my own experience I wouldn’t overrate this, though, since adding whatever decorum is chosen (e.g. a macro) is often done with an editor’s bulk change facilities.

> I also don't agree with this point at all, and I feel that it's purely subjective, and likely based in a lack of familiarity because try blocks are still unstable.

None of the examples above customarily fit on a single line — default and common formatting rules require three lines and added indentation for a block — which is an objective disadvantage compared to `?` syntax. If I’m missing some visual optimisations due to unfamiliarity, would you please add some examples that may reasonably be expected to be stabilised soon?

**Comment by camelid** [comment]
ping from triage: What's the status of this PR? It sounds like maybe it should be closed for now while try blocks are in the works?

**Comment by rkuhn** [comment]
As far as I’m concerned no proposal has been put forward that offers to solve the issue using `try` blocks, so in that sense I don’t agree with the assessment, @camelid.

**Comment by GoldsteinE** [comment]
@m-ou-se I’m not sure what the status of it is. It feels rejected, but there was no formal rejection from libs team, so I’m kinda confused. What’s the verdict on this is?

**Comment by yaahc** [comment]
> @m-ou-se I’m not sure what the status of it is. It feels rejected, but there was no formal rejection from libs team, so I’m kinda confused. What’s the verdict on this?

not currently a libs team member and I'm not recently in the loop on this, but I can clarify the stance I was holding last time I commented on this PR.

Given that this is an irreversible decision I think we need to maximize the amount of information we have before moving forward with this. I raised a blocking objection that is conditional upon `try` blocks being stabilized. Once we have them and have more information I think we could reconsider this issue and it may be justified, but until that time I think the uncertainty, risk, and potential cost are too high.

**Comment by GoldsteinE** [comment]
I still don't see a good way to solve the issue with `try` blocks. What information would we get from stabilization that we don't have now?

**Comment by joshtriplett** [comment]
By way of one example, if we had iterators and try, it'd be feasible to write `yield try { ... expr()? ... }` as a construct to return `Some(Err(...))`. That seems clearer and more explicit than just `expr()?`.

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/libs-api meeting.

We were generally in agreement with @scottmcm's comment at https://github.com/rust-lang/rust/pull/99333#issuecomment-1186263450.

We did look at the code samples for code that would be shortened by these `?` impls, but felt that even those examples did not seem sufficient to add this additional complexity.

We're hoping that some combination of `try` and `gen` and the existing `?` will be able to simplify those use cases. But even if there are cases that those do not handle, we still felt that having this `?` impl would make subsequent *reading* of the code less clear.

@rfcbot close

**Comment by GoldsteinE** [comment]
Thanks for clearing up the status on this PR. I hope generators will remove the `Iterator<Item = Result<_, _>>` usecase sometime in the future. Sadly, `.filter_map()` usecase is not solved by this, so will continue to require a manual `match` on every fallible function call (unless maybe some version of “alternative `try`” is accepted + `try { try {`?).

**Comment by traviscross** [comment]
+1 on **close** and to the concerns that @scottmcm raised [here](https://github.com/rust-lang/rust/pull/99333#issuecomment-1186263450).  In particular, we should add this impl *only if* we're also willing to support "`?` on an `Option<Result<T, E>>` in a function returning `Result<T, E>`". 

Perhaps it goes without saying, but when this is needed, note that macros are always an option.  E.g., we can of course write:

```rust
macro_rules! otry {
    ($e:expr) => { match $e {
        Err(e) => return Some(Err(e)),
        Ok(x) => x,
    }}
}

use core::num::TryFromIntError;
let array: [Result<i16, TryFromIntError>; 4] =
    [Ok(1), Ok(-256), Ok(3), Ok(4)];
let mut iter = array.into_iter().filter_map(|rslt| {
    let _u8 = otry!(u8::try_from(otry!(rslt)));
    (_u8 % 2 == 0).then_some(Ok::<_, TryFromIntError>(_u8))
});
assert!(matches!(iter.next().unwrap(), Err(_)));
assert!(matches!(iter.next().unwrap(), Ok(4)));
```

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&code=fn+main%28%29+%7B%0A++++macro_rules%21+otry+%7B%0A++++++++%28%24e%3Aexpr%29+%3D%3E+%7B+match+%24e+%7B%0A++++++++++++Err%28e%29+%3D%3E+return+Some%28Err%28e%29%29%2C%0A++++++++++++Ok%28x%29+%3D%3E+x%2C%0A++++++++%7D%7D%0A++++%7D%0A%0A++++use+core%3A%3Anum%3A%3ATryFromIntError%3B%0A++++let+array%3A+%5BResult%3Ci16%2C+TryFromIntError%3E%3B+4%5D+%3D%0A++++++++%5BOk%281%29%2C+Ok%28-256%29%2C+Ok%283%29%2C+Ok%284%29%5D%3B%0A++++let+mut+iter+%3D+array.into_iter%28%29.filter_map%28%7Crslt%7C+%7B%0A++++++++let+_u8+%3D+otry%21%28u8%3A%3Atry_from%28otry%21%28rslt%29%29%29%3B%0A++++++++%28_u8+%25+2+%3D%3D+0%29.then_some%28Ok%3A%3A%3C_%2C+TryFromIntError%3E%28_u8%29%29%0A++++%7D%29%3B%0A++++assert%21%28matches%21%28iter.next%28%29.unwrap%28%29%2C+Err%28_%29%29%29%3B%0A++++assert%21%28matches%21%28iter.next%28%29.unwrap%28%29%2C+Ok%284%29%29%29%3B%0A%7D%0A)


That has the same shape as the desired code in [this comment](https://github.com/rust-lang/rust/pull/99333#issuecomment-1186277767).

**Comment by the8472** [comment]
> Perhaps it goes without saying, but when this is needed, note that macros are always an option.

And with [postfix macros](https://github.com/rust-lang/rfcs/pull/2442) it could exist in the same position as `?`

**Comment by GoldsteinE** [comment]
> macros are always an option

Yeah, but macros kill autocompletion.

I’m more hopeful about the `gen`/`try` approach:

```rust
use core::num::TryFromIntError;
let array: [Result<i16, TryFromIntError>; 4] =
    [Ok(1), Ok(-256), Ok(3), Ok(4)];
let mut iter = gen {
    for rslt in array {
        let _u8 = try { u8::try_from(rslt?)? };
        if _u8.is_ok_and(|x| x % 2 == 0) {
            yield Ok(_u8);
        }
    }
});
assert!(matches!(iter.next().unwrap(), Err(_)));
assert!(matches!(iter.next().unwrap(), Ok(4)));
```

In simple cases like this you could just replace `.filter_map()` with an inline generator. It doesn’t exactly generalize to more complex cases (notably, it won’t work with rayon’s parallel iterators, would require `async gen` to work with streams and won’t work with lending iterators at all), but it handles the most common case without compromising on IDE experience.
